src/libraries/Microsoft.Extensions.DependencyInjection.Abstractions/src/ServiceCollection.cs

```diff
@@ -81,13 +103,116 @@ public int IndexOf(ServiceDescriptor item)
         /// <inheritdoc />
         public void Insert(int index, ServiceDescriptor item)
         {
+            // In theory, we should preserve the relative order here but it's expensive to find that out
+            // and we handle operations that rely on the relative order already
+            AddToServiceTypeMap(item);
+
             _descriptors.Insert(index, item);
         }
 
         /// <inheritdoc />
         public void RemoveAt(int index)
         {
+            // This will throw if the index is out of range
+            ServiceDescriptor item = _descriptors[index];
+            if (_serviceTypes.TryGetValue(item.ServiceType, out List<ServiceDescriptor> items))
+            {
+                items.Remove(item);
+            }
+
             _descriptors.RemoveAt(index);
         }
+
+        internal bool TryAdd(ServiceDescriptor item)
```

Comment:
I thought these were going to be public... That's why we took this to API review.

Reply:
Yea, I know but now I'm not sure it needs to be.

Reply:
If we just want to rely on the extension methods... I'm fine with that. But we should go back to the API issue and update it that we only moved the type and didn't add the APIs.

Reply:
Base class doesn't solve any problems; the issue is if someone adds the interface `IServiceCollection` to a class with an existing heirachy introducing a base class doesn't give you anything?

Reply:
Base class doesn't solve any problems; the issue is if someone has the interface `IServiceCollection` on a class with an existing heirachy introducing a base class doesn't give you anything?

They can't now inherit from the base class as C# doesn't support multiple inheritance. (Which is where DIMs do work as they can override the DIM methods)

Reply:
> If we just want to rely on the extension methods... I'm fine with that. But we should go back to the API issue and update it that we only moved the type and didn't add the APIs.

Can do. This came up in the review, all of the APIs are built on `IServiceCollection` so there needs to be a cast anyways.

@benaadams We decided we don't like DIMs in extensions because of the ns2.0 debacle. We would rather add a base class and make people rebase that want these optimizations (which would be another issue).

Reply:
> We would rather add a base class and make people rebase that want these optimizations

Which is essentially forcing breaking changes on the implementors of the interface?

Also how something like `Cysharp/MessagePipe` [`struct ContainerBuilderProxy : IServiceCollection`](https://github.com/Cysharp/MessagePipe/blob/master/src/MessagePipe.Unity/Assets/Plugins/MessagePipe.VContainer/Runtime/TypeProxy.cs) is a struct so can't inherit from a base class.

Just keeping to the dotnet org `dotnet/maui` already has serveral implementations of `IServiceCollection` and is exposing multiple derived `IServiceCollections` as part of its public api e.g. `IMauiServiceCollection` looks to be used as part of hot reload in some way https://github.com/dotnet/maui/pull/851

Reply:
> Which is essentially forcing breaking changes on the implementors of the interface?

It's not a breaking change. The implementors can continue doing what they are doing today, they just don't get this optimization.

> something like Cysharp/MessagePipe struct ContainerBuilderProxy : IServiceCollection is a struct

Isn't that a bad idea because you are boxing every time you pass it into a method that takes `IServiceCollection`?

Reply:
YEP 😄

Reply:
Also I assume if you include a ns2.0 lib that has a transient dependency with a net6 vairant and use it in a net6 app it will use the net6 version? So not sure what the issue is there? 

Cross compiling `Microsoft.Extensions.DependencyInjection.Abstractions` with DIMs will just mean it doesn't have a fast path for the DIMs on Framework? Upgrade to .NET6+; a newer classic Mono or an [updated Unity](https://forum.unity.com/threads/unity-future-net-development-status.1092205/) and its will be good. Don't think its a probablem that only Framework doesn't have the option of a faster path with DIMs? 

They aren't taking anything away, can still rebase on `ServiceCollection` (and then have issues that can't override any of the methods to fit with custom changes). DIMs keep the implementation open, whereas base class closes routes of inhetience and doesn't ingetrate unless everything is virtual, but then that's not great either.

It would be less of an issue if `IServiceCollection` was new and no-one had implemented it but is been out there for 5 years, so saying everyone that implements `IServiceCollection` should inherit from `ServiceCollection` seems a bit much? Also kinda makes the interface pointless?

Reply:
I think we should open a new issue about extensibility. This issue is about optimizing the 90% case. I'm not a fan of using DIMs in libraries that cross compile, it feels wrong to differ the contract based on runtime capabilities. That said, it's unrelated to optimizing this code path.

Reply:
> > Which is essentially forcing breaking changes on the implementors of the interface?
> 
> It's not a breaking change. The implementors can continue doing what they are doing today, they just don't get this optimization.

Which is why I don't understand how adding some DIMs would be an issue? It wouldn't effect anyone that doesn't want to implement them (they'd still go via this path); while giving the option to optimize. However I don't see how a base class would add anything other than making all the methods virtual for everybody?

Reply:
> I'm not a fan of using DIMs in libraries that cross compile, it feels wrong to differ the contract based on runtime capabilities.

It falls out as estension methods where no runtime support and overridable DIM methods for runtimes with support. 

Don't see how is much different than the System.Memory package offerring extensions for string etc to add `.AsSpan()/.AsMemory()` which exist as instance methods when the runtime has support.

Reply:
Span and Memory aren't abstractions that you implement, that's why its different. I don't like the fact that contracts are different across TFMs, that's what feels wrong.